Add errors if we try to use PyTorch with Python 2

dreiss opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #35625 Remove structseq_slice
* #35614 Remove (most) Python 2 support from C++ code
* **#35611 Add errors if we try to use PyTorch with Python 2**
* #35627 Remove __future__ imports from many files
* #35624 Remove a use of exec
* #35623 Remove a use of checkScript(str)
* #35622 Remove caffe2/python/compatibility
* #35621 Remove _get_wrapped_func
* #35620 Remove _compatible_subtest
* #35619 Remove _reset_warning_registry
* #35618 Remove _uses_true_division
* #35617 Remove (most) Python 2 support from setup.py
* #35616 Remove (most) Python 2 support from CI scripts
* #35615 Remove (most) Python 2 support from Python code

Summary:
Python 2 has reached end-of-life and is no longer supported by PyTorch.
To avoid confusing behavior when trying to use PyTorch with Python 2,
detect this case early and fail with a clear message.  This commit
covers "import torch" and attempting to build torch/csrc.

Test Plan:
- Attempted to "import torch" with Python 2 and saw a clear error.
- Attempted to build PyTorch with Python 2 and saw a clear error.

Differential Revision: [D20842878](https://our.internmc.facebook.com/intern/diff/D20842878)

dreiss commented:
@soumith , I'm confused by this failure.  Do you know if `binary_linux_libtorch_2_7m_cpu_gcc5_4_cxx11-abi_shared-with-deps_build` is supposed to be a Pythonless libtorch build?  If so, do you know why it is building a file that includes `torch/csrc/python_headers.h`?

ezyang commented:
I'm not convinced.

Re (1), the user is more likely to run into a syntax error first when they try to import PyTorch. I'm not sure the runtime exception is buying you all that much. And we can prevent people from trying to install PyTorch on Python 2 by advertising appropriately in setup.py that it's python 3 only.

Re (2), shouldn't cmake check Python version and error out earlier than when you try to compile?

dreiss committed:
Update on "Add errors if we try to use PyTorch with Python 2"


Summary:
Python 2 has reached end-of-life and is no longer supported by PyTorch.
To avoid confusing behavior when trying to use PyTorch with Python 2,
detect this case early and fail with a clear message.  This commit
covers "import torch" and attempting to build torch/csrc.

Test Plan:
- Attempted to "import torch" with Python 2 and saw a clear error.
- Attempted to build PyTorch with Python 2 and saw a clear error.

Differential Revision: [D20842878](https://our.internmc.facebook.com/intern/diff/D20842878)

dreiss committed:
Update on "Add errors if we try to use PyTorch with Python 2"


Summary:
Python 2 has reached end-of-life and is no longer supported by PyTorch.
To avoid confusing behavior when trying to use PyTorch with Python 2,
detect this case early and fail with a clear message.  This commit
covers "import torch" and attempting to build torch/csrc.

Test Plan:
- Attempted to "import torch" with Python 2 and saw a clear error.
- Attempted to build PyTorch with Python 2 and saw a clear error.

Differential Revision: [D20842878](https://our.internmc.facebook.com/intern/diff/D20842878)

dreiss committed:
Update on "Add errors if we try to use PyTorch with Python 2"


Summary:
Python 2 has reached end-of-life and is no longer supported by PyTorch.
To avoid confusing behavior when trying to use PyTorch with Python 2,
detect this case early and fail with a clear message.  This commit
covers "import torch" and attempting to build torch/csrc.

Test Plan:
- Attempted to "import torch" with Python 2 and saw a clear error.
- Attempted to build PyTorch with Python 2 and saw a clear error.

Differential Revision: [D20842878](https://our.internmc.facebook.com/intern/diff/D20842878)

[ghstack-poisoned]

dreiss committed:
Update on "Add errors if we try to use PyTorch with Python 2"


Summary:
Python 2 has reached end-of-life and is no longer supported by PyTorch.
To avoid confusing behavior when trying to use PyTorch with Python 2,
detect this case early and fail with a clear message.  This commit
covers "import torch" and attempting to build torch/csrc.

Test Plan:
- Attempted to "import torch" with Python 2 and saw a clear error.
- Attempted to build PyTorch with Python 2 and saw a clear error.

Differential Revision: [D20842878](https://our.internmc.facebook.com/intern/diff/D20842878)

[ghstack-poisoned]

dreiss committed:
Update on "Add errors if we try to use PyTorch with Python 2"


Summary:
Python 2 has reached end-of-life and is no longer supported by PyTorch.
To avoid confusing behavior when trying to use PyTorch with Python 2,
detect this case early and fail with a clear message.  This commit
covers "import torch" and attempting to build torch/csrc.

Test Plan:
- Attempted to "import torch" with Python 2 and saw a clear error.
- Attempted to build PyTorch with Python 2 and saw a clear error.

Differential Revision: [D20842878](https://our.internmc.facebook.com/intern/diff/D20842878)

[ghstack-poisoned]
